tasmota/tasmota_xsns_sensor/xsns_115_wooliis.ino

```diff
@@ -0,0 +1,230 @@
+    AddLogBuffer(LOG_LEVEL_DEBUG_MORE, buffer, 21);
+    return;
+  }
+ 
+  uint8_t checksum = 0;
+  for (uint32_t i = 0; i < 20; i++) {
+    checksum += buffer[i];
+  }
+  checksum = 255 - checksum;
+  if (checksum != buffer[20]) {
+    AddLog(LOG_LEVEL_DEBUG, PSTR("Wooliis: " D_CHECKSUM_FAILURE));
+    return;
+  }
+
+  AddLog(LOG_LEVEL_DEBUG, PSTR("Wooliis: message received correctly!"));
+  AddLogBuffer(LOG_LEVEL_DEBUG_MORE, buffer, 21);
+
+  Wooliis.charge_percent = buffer[4];
+  Wooliis.remaining_capacity = 0.1f * ((uint16_t)buffer[5]<<8 | buffer[6]);
+  Wooliis.voltage = 0.1f * ((uint16_t)buffer[7]<<8 | buffer[8]);
+  Wooliis.current = (buffer[19] ? -0.1f : 0.1f) * ((uint16_t)buffer[9]<<8 | buffer[10]);
+  Wooliis.energy_in = 0.1f * ((uint32_t)buffer[11]<<16 | (uint16_t)buffer[12]<<8 | buffer[13]);
+  Wooliis.energy_out = 0.1f * ((uint32_t)buffer[14]<<16 | (uint16_t)buffer[15]<<8 | buffer[16]);
+  Wooliis.status = buffer[19];
+
+  Wooliis.valid = 1;
+
+  return;
+}
+
+/*********************************************************************************************/
+
+void WooliisSecond(void)                 // Every second
+{
+  if (Wooliis.ready)
+    WooliisReadData();
+}
+
+/*********************************************************************************************/
+
+void WooliisInit(void) {
+  if (PinUsed(GPIO_WOOLIIS_RX)) {
+    WooliisSerial = new TasmotaSerial(Pin(GPIO_WOOLIIS_RX), -1, 1);
+    if (WooliisSerial->begin(9600)) {
+      if (WooliisSerial->hardwareSerial()) { ClaimSerial(); }
+      WooliisSerial->flush();  // Clear serial buffer
+      Wooliis.ready = 1;
+      AddLog(LOG_LEVEL_DEBUG, PSTR("Wooliis: Serial ready"));
+    }
+  }
+  Wooliis.valid = 0;
+}
+
+#ifdef USE_WEBSERVER
+const char HTTP_SNS_WOOLIIS_BCP[] = "{s}%s " D_BATTERY_CHARGE "{m}%d " D_UNIT_PERCENT "{e}";
+const char HTTP_SNS_WOOLIIS_CAP[] = "{s}%s " D_CAPACITY "{m}%s " D_UNIT_CHARGE "{e}";
+const char HTTP_SNS_WOOLIIS_CHG[] = "{s}%s " D_CHARGING "{m}%s{e}";
+const char HTTP_SNS_WOOLIIS_VOL[] = "{s}%s " D_VOLTAGE "{m}%s " D_UNIT_VOLT "{e}";
+const char HTTP_SNS_WOOLIIS_CUR[] = "{s}%s " D_CURRENT "{m}%s " D_UNIT_AMPERE "{e}";
+const char HTTP_SNS_WOOLIIS_POW[] = "{s}%s " D_POWERUSAGE "{m}%s " D_UNIT_WATT "{e}";
+const char HTTP_SNS_WOOLIIS_EIN[] = "{s}%s " D_ENERGY " " D_INGRESS "{m}%s " D_UNIT_WATTHOUR "{e}";
+const char HTTP_SNS_WOOLIIS_EEG[] = "{s}%s " D_ENERGY " " D_EGRESS "{m}%s " D_UNIT_WATTHOUR "{e}";
+#endif  // USE_WEBSERVER
+
+void WooliisShow(bool json) {
+  char float_str[8];
+  if (Wooliis.valid) {
+    if (json) {
+      ResponseAppend_P(PSTR(",\"%s\":{"), D_WOOLIIS);
+      dtostrfd(Wooliis.remaining_capacity, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s,"), D_CAPACITY, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%d,"), D_CHARGING, (Wooliis.status ? 1 : 0));
+      dtostrfd(Wooliis.voltage, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s,"), D_VOLTAGE, float_str);
```

Comment:
Learn to use ``%*_f`` and allow users to change the amount of decimals shown like in xdrv_03_energy.ino using ``Settings->flag2.voltage_resolution``

Reply:
Hi Theo, this is probably the only problematic point. I tried format strings of different sorts, %f %.1f %*_f, and none of them worked (some did not display decimals, some returned a string like "undefined"). I just tried again today as you mentioned and the GPIO config does not survive a reboot after I set the pin to my new sensor. I assume there is some reset kicking in when some crash happen. I would be happy to continue troubleshooting, but I don't have too many other insights on how to proceed. Any recommendation?

Reply:
Yes. Sorry. In order to use ``%*_f`` you'll need to reference the variable by address like:

```
ResponseAppend_P(PSTR("\"" D_JSON_VOLTAGE "\":%*_f"), Settings->flag2.voltage_resolution, &Wooliis.voltage);
```

Notice the "&" before the float variable.

Reply:
Documentation is here https://tasmota.github.io/docs/Sensor-API/#extension-to-vsnprintf